include/itkVkMultiResolutionPyramidImageFilter.hxx

```diff
@@ -0,0 +1,210 @@
+    shrinker = ShrinkerType::New();
+    shrinkerFilter = shrinker.GetPointer();
+  }
+  else
+  {
+    resampleShrinker = ResampleShrinkerType::New();
+    using LinearInterpolatorType = itk::LinearInterpolateImageFunction<OutputImageType, double>;
+    auto interpolator = LinearInterpolatorType::New();
+    resampleShrinker->SetInterpolator(interpolator);
+    resampleShrinker->SetDefaultPixelValue(0);
+    shrinkerFilter = resampleShrinker.GetPointer();
+  }
+  // Setup the filters
+  caster->SetInput(inputPtr);
+
+  // Set up smoothing filters
+  spatialSmoother->SetUseImageSpacing(false);
+  spatialSmoother->SetInput(caster->GetOutput());
+  spatialSmoother->SetMaximumError(this->m_MaximumError);
+  fftSmoother->SetUseImageSpacing(false);
+  fftSmoother->SetInput(caster->GetOutput());
+  fftSmoother->SetMaximumError(this->m_MaximumError);
+
+  unsigned int ilevel, idim;
+  unsigned int factors[ImageDimension];
+  VarianceType variance;
+
+  for (ilevel = 0; ilevel < this->m_NumberOfLevels; ++ilevel)
+  {
+    this->UpdateProgress(static_cast<float>(ilevel) / static_cast<float>(this->m_NumberOfLevels));
+
+    // Allocate memory for each output
+    OutputImagePointer outputPtr = this->GetOutput(ilevel);
+    outputPtr->SetBufferedRegion(outputPtr->GetRequestedRegion());
+    outputPtr->Allocate();
+
+    // compute shrink factors
+    for (idim = 0; idim < ImageDimension; ++idim)
+    {
+      factors[idim] = this->m_Schedule[ilevel][idim];
+    }
+
+    if (!this->GetUseShrinkImageFilter())
+    {
+      using IdentityTransformType = itk::IdentityTransform<double, OutputImageType::ImageDimension>;
+      auto identityTransform = IdentityTransformType::New();
+      resampleShrinker->SetOutputParametersFromImage(outputPtr);
+      resampleShrinker->SetTransform(identityTransform);
+    }
+    else
+    {
+      shrinker->SetShrinkFactors(factors);
+    }
+
+    // select spatial or FFT smoothing based on user threshold settings
+    // to maximize anticipated performance
+    if (GetUseFFT(this->GetKernelRadius(ilevel)))
+    {
+      smoother = fftSmoother;
+    }
+    else
+    {
+      smoother = spatialSmoother;
+    }
+    variance = this->GetVariance(ilevel);
+    smoother->SetVariance(variance);
+    shrinkerFilter->SetInput(smoother->GetOutput());
+
+    shrinkerFilter->GraftOutput(outputPtr);
+
+    // force to always update in case shrink factors are the same
+    shrinkerFilter->Modified();
+    shrinkerFilter->UpdateLargestPossibleRegion();
+    this->GraftNthOutput(ilevel, shrinkerFilter->GetOutput());
+  }
+}
+
+template <typename TInputImage, typename TOutputImage>
+float
+VkMultiResolutionPyramidImageFilter<TInputImage, TOutputImage>::ComputeMetricValue(
+  const InputSizeType &  inputSize,
+  const KernelSizeType & kernelRadius) const
+{
+  unsigned int totalKernelSize = 0;
+  float        metricValue = 1.0f;
+  for (unsigned int dim = 0; dim < ImageDimension; ++dim)
+  {
+    totalKernelSize += kernelRadius[dim] * 2 + 1;
```

Comment:
It appears that we are using, more-or-less, `(I^2 + J^2 + K^2)` rather than `(I + J + K)` in our metric formula.  If so and if that is what we want, we should change the documentation accordingly.  However, see the note below.

Reply:
Never mind.  I saw multiplication by 2 and thought it was squaring.